src/bpf/bpfctl.c

```diff
@@ -0,0 +1,220 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <getopt.h>
+
+#include "build.h"
+#include "format-table.h"
+#include "format-util.h"
+#include "main-func.h"
+#include "parse-argument.h"
+#include "pretty-print.h"
+#include "varlink.h"
+#include "verbs.h"
+
+static JsonFormatFlags arg_json_format_flags = JSON_FORMAT_OFF;
+
+struct bpf_prog_data {
+        char *name;
+        uint64_t type;
+        char *memlock;
+        char *tag;
+        char *map_names;
+        uint64_t run_time_ns;
+        uint64_t run_cnt;
+};
+
+struct bpf_map_data {
+        char *name;
+        uint64_t type;
+        char *memlock;
+};
+
+static int query_bpf_progs_and_maps_reply(
+                Varlink *link,
+                JsonVariant *parameters,
+                const char *error_id,
+                VarlinkReplyFlags flags,
+                void *userdata) {
+
+        int r;
+        JsonVariant *m, *v;
+
+        // json_variant_dump(parameters, JSON_FORMAT_PRETTY, NULL, NULL);
+
+        m = json_variant_by_key(parameters, "TYPE");
+        if (!m)
+                return 0;
+        v = json_variant_by_key(parameters, "DATA");
+        if (!v)
+                return 0;
+
+        if (streq(json_variant_string(m), "MAP")) {
+                static const JsonDispatch dispatch_table[] = {
+                        { "NAME",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_map_data, name),  JSON_MANDATORY },
+                        { "TYPE", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct bpf_map_data, type), JSON_MANDATORY},
+                        { "MEMLOCK", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_map_data, memlock), JSON_MANDATORY},
+                        {}
+                };
+                struct bpf_map_data p = {};
+
+                r = json_dispatch(v, dispatch_table, NULL, 0, &p);
+                if (r < 0)
+                        return r;
```

Comment:
this would leak the strings allocated on the heap that are stored in `bpf_map_data`.

please add a destructor call `bpf_map_data_done()` that frees the fields of the structure. Then use `_cleanup_(bpf_map_data_done() struct bpf_map_data p = {};` to allocate the struct and free it automatically when it goes out of scope.